stack/building-with-pip.rst

```diff
@@ -0,0 +1,241 @@
+There are checkers available that can block merging if such a change has been made.
+
+If you want the version number of the build to be determined correctly the code must be checked out on GitHub with the full history included:
+
+.. code-block:: yaml
+
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          # Need to clone everything for the git tags.
+          fetch-depth: 0
+
+Once a package is pip-installable the package can be tested in the GitHub action.
+If ``pytest`` is configured with code coverage enabled the results can be uploaded to CodeCov and reported on the pull request.
+This would look something like:
+
+.. code-block:: yaml
+
+    - name: Build and install
+      run: |
+        python -m pip install --no-deps -v -e .
+
+    - name: Run tests
+      run: |
+        pytest -r a -v -n 3 --open-files --cov=lsst.resources\
+                --cov=tests --cov-report=xml --cov-report=term --cov-branch
+
+    - name: Upload coverage to codecov
+      uses: codecov/codecov-action@v2
+      with:
+        file: ./coverage.xml
+
+
+
+Distributing the Package on PyPI
+================================
+
+Once the package supports ``pip install`` it is a small configuration change to allow it to be distributed on `PyPI`_.
+One caveat is that all the required dependencies listed in the ``pyproject.toml`` file must exist on `PyPI`_.
+
+The recommended process is for a GitHub action to trigger when the package is tagged.
+This action will then build the package and trigger the upload to `PyPI`_.
+All science pipeline packages on `PyPI`_ must be owned by the Rubin DM `PyPI`_ account attached to ``dm-admin@lists.lsst.org``.
+
+The `PyPI`_ upload can be configured in the same GitHub action that builds the package and tests it.
+Usually it will block on the successful completion of that phase and then only trigger if a tag is being added.
+
+A full example can be seen below:
+
+.. code-block:: yaml
+
+  pypi:
+
+    runs-on: ubuntu-latest
+    needs: [build_and_test]
+    if: startsWith(github.ref, 'refs/tags/')
+    permissions:
+      id-token: write
+
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          # Need to clone everything to embed the version.
+          fetch-depth: 0
+
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.11"
+
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install --upgrade setuptools wheel build
+
+      - name: Build and create distribution
+        run: |
+          python -m build --skip-dependency-check
+
+      - name: Upload
+        uses: pypa/gh-action-pypi-publish@release/v1
+
+For the upload to work `PyPI`_ must be preconfigured to expect uploads from this specific GitHub action using a `trusted publisher`_ mechanism.
```

Comment:
Who does this?  How does the author request it?

Reply:
I was wondering myself how you can warn pypi that a package called lsst-x is about to turn up.